### DISCUSSION TITLE
想要给这个程序增加输入法功能有没有可能？

### DISCUSSION THREAD

**Discussion by wastee** [fcitx/fcitx5#518] [Q&A] [answered]
## 需求

我想给 [REAPER](http://reaper.fm/) 这个软件增加输入法功能

### 这个软件的介绍

1. 这个软件是跨平台的、闭源的
2. 这个软件的 Linux 图形是由 Windows “翻译”过来的，这个“翻译”是开源的 
    https://github.com/justinfrankel/WDL/blob/main/WDL/swell/swell-generic-gdk.cpp

### 我目前基本不了解的情况

1. 我没有 C++ 基础
2. 我不懂 GDK 和 GTK 有什么关系
3. 我不能确定这个软件是否在 Linux 下是“翻译”为 GDK
4. 我不能确定 GDK 软件有没有可能支持 fcitx5 输入法

### 我目前稍微了解的

- 有 Linux 用户修改 [WDL/swell](https://github.com/justinfrankel/WDL/pull/12/files)，来达到修改界面的目的

## 我想知道的

1. 这个软件的“[翻译](https://github.com/justinfrankel/WDL/tree/main/WDL/swell)”层源码里，既有 GDK 又有 GTK。那这个软件究竟是 GDK 还是 GTK的？
2. **这个软件有没有可能支持输入法功能**
3. 这个软件是否与 sublime-text 的情况类似，即跟这个 [patch](https://github.com/lyfeyaj/sublime-text-imfix) 相像？

------

此前在软件官方论坛提过功能请求与咨询，开发者并没有回复，该软件官方论坛也没有更多使用输入法的 Linux 用户了解上述需求可能性。

在 C++/GDK/GTK，我当前只有本贴的需求。学习前我最最关心的，就是有没有可能，于是在此咨询各位，再去出发。

谢谢各位的时间！

**Comment by wengxt** [comment, marked as answer]
GDK 是 GTK 底层的只和图形绘制有关系的 API，GTK 是用 GDK 实现的，输入法的实现都在 gtk 的 wdiget 里

我大概读了一下 swell 的代码，看起来是用 gdk 的来模拟 win32 的部分，假定你能替换里面的 libSwell 成为你的版本
那么你需要做的就是
1、给 libswell 开启 Gtk 支持，主要指 SWELL_SUPPORT_GTK=1
2、创建一个 GtkIMContext 对象 https://developer-old.gnome.org/gtk3/stable/GtkIMContext.html
3、和 GtkIMContext 对象交互（例如把按键交给它处理 https://developer-old.gnome.org/gtk3/stable/GtkIMContext.html#gtk-im-context-filter-keypress 处理程序的焦点），并且把对应事件（例如  https://developer-old.gnome.org/gtk3/stable/GtkIMContext.html#GtkIMContext-commit ）翻译成对应的 Windows 的输入法事件 https://docs.microsoft.com/en-us/windows/win32/intl/wm-ime-char 

和 sublime 的情况完全不象，sublime 的控件和窗口都是 Gtk 的，只是在具体和输入法相关的 API 上有一些小问题，缺少一些调用

**Reply by wastee** [reply]
很清晰易懂，让我少饶非常多弯路！非常感谢！
